### PR TITLE
Adding serving function for tf_records with unknown batch size.

### PR DESCRIPTION
**Pull request by fprost**
Please review this.

There is one "ugly" workaround. 
The tokenizer is different for training and for serving. Indeed, for training the tokenizer will be used within a map function and then processes one example at a time, whereas for serving the tokenizer needs to process an entire batch.

Therefore, in run.py, I initialize twice the tokenizer. It does not seem a big deal to me but I acknowledge it could be improved. I tried to merge (by doing all per batch or per example) but unsuccessfully.
Might be a TODO or a github issue.

Let me know your thoughts.

### CODE REVIEW

File: experiments/tf_trainer/common/serving_input.py [hunk 1]
```diff
@@ -7,9 +7,15 @@
 import tensorflow as tf
 from tensorflow.python.ops import array_ops
 
+FLAGS = tf.app.flags.FLAGS
+
+tf.app.flags.DEFINE_boolean("use_tfrecords_for_serving", True,
```
[the remark this answers is not in the document]
Comment (fprost) [reply experiments/tf_trainer/common/serving_input.py]: Done.
Do you know if there is a way to do a list of possible answers with tf.app.flags?
I put an exception in the code but would be nicer to put in flags.

File: experiments/tf_trainer/common/text_preprocessor.py [hunk 2]
```diff
@@ -62,15 +62,21 @@ def _tokenize_tensor_op(text: types.Tensor) -> types.Tensor:
...
           words_int_sparse.indices,
           words_int_sparse.dense_shape,
           words_int_sparse.values,
           default_value=0)
 
-      return tf.squeeze(words_int_dense)
+      if single_record_level:
+        return tf.squeeze(words_int_dense)
+      else:
+        return words_int_dense
```
[the remark this answers is not in the document]
Comment (fprost) [reply experiments/tf_trainer/common/text_preprocessor.py]: Done.

File: experiments/tf_trainer/keras_gru_attention/run.py [hunk 3]
```diff
@@ -32,7 +32,7 @@
                            "required for serving.")
 tf.app.flags.DEFINE_integer("batch_size", 64,
                             "The batch size to use during training.")
-tf.app.flags.DEFINE_integer("train_steps", 100,
+tf.app.flags.DEFINE_integer("train_steps", 0,
```
[the remark this answers is not in the document]
Comment (fprost) [reply experiments/tf_trainer/keras_gru_attention/run.py]: Should not have been part of the commit. 
Changed back to 1000.

File: experiments/tf_trainer/common/serving_input.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nthain) [reply experiments/tf_trainer/common/serving_input.py]: I don't know of any way to force it. I think adding the list in the help string and the assert/exception should be fine. Thanks!

File: experiments/tf_trainer/common/text_preprocessor.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nthain) [reply experiments/tf_trainer/common/text_preprocessor.py]: I defer to your judgement. I think it's better safe than sorry, so I'm happy to keep it explicit as is.

File: experiments/tf_trainer/keras_gru_attention/run.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nthain) [reply experiments/tf_trainer/keras_gru_attention/run.py]: Thanks!

**Comment by fprost** [comment]
Actually cleaned the preprocessing function. It now takes only 1-D string tensor (before it was scalar for training, 1-d for serving).
A few minor tensor operations in training make it compatible!

**Pull request merged**